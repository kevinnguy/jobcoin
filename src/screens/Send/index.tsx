import React, { useState } from 'react';
import { FlatList, Text, View } from 'react-native';
import currency from 'currency.js';

import { Transaction, useAuth } from '../../auth/useAuth';
import { Button } from '../../components/Button';
import { Chart } from '../../components/Chart';
import { SendModal } from './SendModal';

import styles from './styles';

export const SendScreen = () => {
  const auth = useAuth();
  const user = auth.user!;
  const onPressSignOut = () => auth.signOut();

  const [isModalVisible, setIsModalVisible] = useState(false);
  const showModal = () => setIsModalVisible(true);
  const dismissModal = () => setIsModalVisible(false);

  const renderHeader = () => {
    const userAmount = currency(user.balance).format();
    const chartData = user.transactions.reduce<{
      total: number;
      data: number[];
    }>(
      (acc, transaction) => {
        const { fromAddress, amount } = transaction;
        const isReceived = fromAddress !== user.address;
        const numericalAmount = currency(amount).value;

        if (isReceived) {
          acc.total = acc.total + numericalAmount;
        } else {
          acc.total = acc.total - numericalAmount;
        }

        acc.data.push(acc.total);
        return acc;
      },
      { total: 0, data: [] },
    );

    return (
      <>
        <View style={styles.header}>
          <Text style={styles.userAmount}>{userAmount}</Text>
          <Button
            style={styles.signOut}
            title={'Sign out'}
            onPress={onPressSignOut}
          />
        </View>
        <Chart data={chartData.data} />
      </>
    );
  };

  const renderItem = ({ item }: { item: Transaction }) => {
    const { fromAddress, toAddress, timestamp, amount } = item;
    const isReceived = fromAddress !== user.address;

    return (
      <View style={styles.transaction}>
        <View>
          <Text style={styles.transactionTitle}>{toAddress}</Text>
          <Text style={styles.transactionDetail}>{timestamp}</Text>
        </View>
        <Text
          style={[
            styles.transactionAmount,
            isReceived ? styles.transactionReceived : styles.transactionSent,
          ]}>{`${isReceived ? '+' : '-'} ${currency(amount).format()}`}</Text>
      </View>
    );
  };

  const renderSeparator = () => <View style={styles.separator} />;

  const transactionsData = [...user.transactions];
  transactionsData.reverse();

  return (
    <View style={styles.container}>
      <SendModal visible={isModalVisible} onRequestClose={dismissModal} />
      <FlatList
        data={transactionsData}
        ListHeaderComponent={renderHeader}
        ItemSeparatorComponent={renderSeparator}
        renderItem={renderItem}
      />
      <View style={styles.inputBar}>
        <Button title={'Send Jobcoin'} onPress={showModal} />
      </View>
    </View>
  );
};
