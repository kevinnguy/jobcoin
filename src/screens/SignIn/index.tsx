import React, { useState } from 'react';
import { Alert, View } from 'react-native';

import { useAuth } from '../../auth/useAuth';
import { Button } from '../../components/Button';
import { Input } from '../../components/Input';

import styles from './styles';

export const SignInScreen = () => {
  const auth = useAuth();
  const [address, setAddress] = useState<string>('');

  const onPressSignIn = () => {
    if (address.length === 0) {
      Alert.alert('Please enter an address to sign in.');
      return;
    }

    auth.signIn(address);
  };

  return (
    <View style={styles.container}>
      <Input
        style={styles.textInput}
        onChangeText={setAddress}
        value={address}
        placeholder={'Enter your address'}
      />
      <Button title={'Sign in'} onPress={onPressSignIn} />
    </View>
  );
};
