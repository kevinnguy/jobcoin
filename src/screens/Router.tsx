import React from 'react';

import { SignInScreen } from './SignIn';
import { SendScreen } from './Send';

import { useAuth } from '../auth/useAuth';

/**
 * If we had more navigation states and screens, it's best to use a library
 * like react-navigation or react-native-navigation
 */
export const Router = () => {
  const auth = useAuth();

  if (auth.user) {
    return <SendScreen />;
  }

  return <SignInScreen />;
};
