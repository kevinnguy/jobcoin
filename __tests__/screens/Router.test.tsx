/* eslint-disable no-labels */
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

// jest-fetch-mock could be used
// @testing-library/react-native can be used

import { Router } from '../../src/screens/Router';

jest.mock('react-native-chart-kit', () => {
  LineChart: () => {};
});

describe('SignIn screen', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders sign in screen if no user', () => {
    jest.mock('../../src/auth/useAuth.tsx', () => {
      useAuth: () => ({
        user: undefined,
      });
    });

    const component = renderer.create(<Router />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders send screen if user exists', () => {
    jest.mock('../../src/auth/useAuth.tsx', () => {
      useAuth: () => ({
        user: 'abc123',
      });
    });

    const component = renderer.create(<Router />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
