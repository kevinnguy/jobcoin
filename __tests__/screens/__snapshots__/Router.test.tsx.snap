// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn screen renders send screen if user exists 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 15,
    }
  }
>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="Enter your address"
    style={
      Array [
        Object {
          "borderColor": "gray",
          "borderRadius": 8,
          "borderWidth": 1,
          "fontSize": 16,
          "height": 50,
          "paddingLeft": 10,
        },
        Object {
          "marginBottom": 15,
        },
      ]
    }
    value=""
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Sign in
    </Text>
  </View>
</View>
`;

exports[`SignIn screen renders sign in screen if no user 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 15,
    }
  }
>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="Enter your address"
    style={
      Array [
        Object {
          "borderColor": "gray",
          "borderRadius": 8,
          "borderWidth": 1,
          "fontSize": 16,
          "height": 50,
          "paddingLeft": 10,
        },
        Object {
          "marginBottom": 15,
        },
      ]
    }
    value=""
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderRadius": 8,
          "height": 50,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Sign in
    </Text>
  </View>
</View>
`;
